### PR TITLE
HPCC-26327 MySQL JSON support

### PR DESCRIPTION
**Pull request by rpastrana**
- Accepts MySQL JSON type as string
- Adds mysql json tests

Signed-off-by: Rodrigo Pastrana <rodrigo.pastrana@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by rpastrana** [comment]
@richardkchapman any concerns about treating mysql json as strings?

**Review by richardkchapman — APPROVED** [review]

**Review by richardkchapman — CHANGES_REQUESTED** [review]
There are build errors on centos7 because it's using an older version of the mysql client:

#54 350.6 /hpcc-dev/HPCC-Platform/plugins/mysql/mysqlembed.cpp:807:10: error: ‘MYSQL_TYPE_JSON’ was not declared in this scope; did you mean ‘MYSQL_TYPE_BLOB’?

Probably need to ad some code to check the client version

**Comment by rpastrana** [comment]
Leaving notes for reference:

centos 7
> yum install -y mysql-devel #16614 (installs mariadb-libs)
> more  /usr/include/mysql/mysql_version.h

#define PROTOCOL_VERSION                10
**#define MYSQL_SERVER_VERSION            "5.5.68-MariaDB"**
#define MYSQL_BASE_VERSION              "mysqld-5.5"
#define MARIADB_BASE_VERSION            "mariadb-5.5"
#define MYSQL_SERVER_SUFFIX_DEF         ""
#define FRM_VER                         6
**#define MYSQL_VERSION_ID                50568**
#define MYSQL_PORT                      3306
#define MYSQL_PORT_DEFAULT              0
#define MYSQL_UNIX_ADDR                 "/var/lib/mysql/mysql.sock"
#define MYSQL_CONFIG_NAME               "my"
#define MYSQL_COMPILATION_COMMENT       "MariaDB Server"


Best clue regarding the introduction version of MYSQL_TYPE_JSON:
https://github.com/MariaDB/server/blob/04aab8283022b9a37969df90f44f0c95b98e01d1/storage/mroonga/mrn_mysql_compat.h#L31

#if MYSQL_VERSION_ID >= 50709 && !defined(MRN_MARIADB_P)
#  define MRN_HAVE_MYSQL_TYPE_JSON
#endif

File: plugins/mysql/mysqlembed.cpp [hunk 1]
```diff
@@ -804,6 +804,9 @@ static bool isString(enum_field_types type)
     case MYSQL_TYPE_BLOB:
     case MYSQL_TYPE_STRING:
     case MYSQL_TYPE_VAR_STRING:
+#if MYSQL_VERSION_ID >= 50709
```
Comment (richardkchapman) [inline plugins/mysql/mysqlembed.cpp:807]: Perhaps you can say

`#ifdef MRN_HAVE_MYSQL_TYPE_JSON
`
?

**Review by richardkchapman — CHANGES_REQUESTED** [review]

File: plugins/mysql/mysqlembed.cpp (hunk 1 above)
Comment (rpastrana) [reply plugins/mysql/mysqlembed.cpp:807]: Considered it, and tested against it, but it seems to be defined outside the client lib.

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**